File: src/physics/Rigidbody.cpp
#include "physics/Rigidbody.h"
#include "math/Vec2.h"

Rigidbody::Rigidbody()
{
	this->velocity = Vec2::zero();
	this->acceleration = Vec2::zero();
	this->mass = 1.0f;
	this->drag = 0.1f;
	this->isStatic = false;
	this->inverseMass = (isStatic || mass == 0.0f) ? 0.0f : 1.0f / mass;
}

Rigidbody::Rigidbody(float mass, float drag, bool isStatic)
{
	this->velocity = Vec2::zero();
	this->acceleration = Vec2::zero();
	this->mass = mass;
	this->drag = drag;
	this->isStatic = isStatic;
	this->inverseMass = (isStatic || mass == 0.0f) ? 0.0f : 1.0f / mass;
}

Rigidbody::~Rigidbody() {}

void Rigidbody::applyForce(const Vec2 &force)
{
	if (isStatic) return;
	acceleration += force * inverseMass;
}

void Rigidbody::applyImpulse(const Vec2 &impulse)
{
	if (isStatic) return;
	velocity += impulse * inverseMass;
}

void Rigidbody::integrate(Vec2 &position, float dt)
{
	if (isStatic) return;

	// update velocity and position
	velocity += acceleration * dt;
	position += velocity * dt;

	// apply drag
	velocity *= (1.0f - drag * dt);

	// reset acceleration for next frame
	acceleration = Vec2(0, 0);
}

File: components/rigidbody.go
package components

import (
	"math"

	. "froglib/math"
)

type Rigidbody struct {
	velocity     Vec2
	acceleration Vec2
	mass         float64
	drag         float64
	isStatic     bool
	inverseMass  float64
}

func NewRigidbody(mass float64, drag float64, isStatic bool) *Rigidbody {
	var inverseMass float64
	if isStatic || mass == 0.0 {
		inverseMass = 0.0
	} else {
		inverseMass = 1.0 / mass
	}

	rigidbody := Rigidbody{
		velocity:     NewVec2(0.0, 0.0),
		acceleration: NewVec2(0.0, 0.0),
		mass:         mass,
		drag:         drag,
		isStatic:     isStatic,
		inverseMass:  inverseMass,
	}

	return &rigidbody
}

func (r *Rigidbody) ApplyForce(force Vec2) {
	if r.isStatic {
		return
	}
	r.acceleration = r.acceleration.Add(force.Scale(r.inverseMass))
}

func (r *Rigidbody) ApplyImpulse(impulse Vec2) {
	if r.isStatic {
		return
	}
	r.velocity = r.velocity.Add(impulse.Scale(r.inverseMass))
}

func (r *Rigidbody) Integrate(position *Vec2) {
	if r.isStatic {
		return
	}

	// update velocity with acceleration
	r.velocity = r.velocity.Add(r.acceleration)

	// apply drag
	r.velocity = r.velocity.Scale(math.Exp(-r.drag))

	// update position with velocity
	*position = position.Add(r.velocity)

	// reset acceleration for next frame
	r.acceleration = NewVec2(0.0, 0.0)
}
